pkg/fqdn/re/re.go

```diff
@@ -0,0 +1,82 @@
+// SPDX-License-Identifier: Apache-2.0
+// Copyright Authors of Cilium
+
+// Package re provides a simple function to access compile regex objects for
+// the FQDN subsystem.
+package re
+
+import (
+	"errors"
+	"fmt"
+	"regexp"
+	"sync/atomic"
+	"unsafe"
+
+	lru "github.com/golang/groupcache/lru"
+
+	"github.com/cilium/cilium/pkg/lock"
+	"github.com/cilium/cilium/pkg/logging"
+	"github.com/cilium/cilium/pkg/logging/logfields"
+	"github.com/cilium/cilium/pkg/option"
+)
+
+var (
+	log = logging.DefaultLogger.WithField(logfields.LogSubsys, "fqdn/re")
+)
+
+// CompileRegex compiles a pattern p into a regex and returns the regex object.
+// The regex object will be cached by an LRU. If p has already been compiled
+// and cached, this function will return the cached regex object. If not
+// already cached, it will compile p into a regex object and cache it in the
+// LRU. This function will return an error if the LRU has not already been
+// initialized.
+func CompileRegex(p string) (*regexp.Regexp, error) {
+	lru := (*RegexCompileLRU)(atomic.LoadPointer(&regexCompileLRU))
+	if lru == nil {
+		return nil, errors.New("FQDN regex compilation LRU not yet initialized")
+	}
+	lru.Lock()
```

Comment:
Why not use a `sync.RWMutex` here? I saw another person made the same suggestion, but the suggestion was marked as resolved without comment or code change.

Reply:
See https://github.com/cilium/cilium/pull/19632#discussion_r869385577.